### PR TITLE
CRTP: Cross Validation

### PR DESCRIPTION
**Pull request by akleeman**
This change (re)introduces cross validation with the ability to specialize for different model implementations.

It's a large change, but I broke it up into smaller commits, probably best to step through those.  One of the later commits contains a lot of tests, those might serve as good examples of how cross validation works.

The result is the ability to do things that would look like this:
```
auto model = make_some_model();
auto dataset = make_some_dataset();
auto leave_one_out_prediction = model.cross_validate().get_prediction(dataset).mean();
```
Specialization of cross validated methods is probably best seen by example (in `gp.h`) but basically by defining a method that looks like:
```
  template <typename FeatureType>
  std::vector<MarginalDistribution>
  cross_validated_predictions(const RegressionDataset<FeatureType> &dataset,
                              const FoldIndexer &fold_indexer,
                              PredictTypeIdentity<MarginalDistribution>) const
```
a model can provide specialized predict methods.

### CODE REVIEW

File: albatross/evaluation/cross_validation_utils.h [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    predictions.emplace_back(model.get_fit_model(fold.train_dataset)
+                                 .get_prediction(fold.test_dataset.features));
+  }
+
+  return predictions;
+}
+
+template <typename FeatureType>
+inline std::size_t
+max_index(const std::vector<RegressionFold<FeatureType>> &folds) {
```
Comment (seth-swiftnav) [inline albatross/evaluation/cross_validation_utils.h]: This function and it's use seems a bit hacky. The function finds the highest index in the folds, but contrary to the name, adds 1 to return the number of original features. There's also an assumption that the highest index appears in the fold's test sets, which is true is using a GrouperFunction, but not true of folds themselves.
It's like there should be two types of folds, a generic one and a strict one where the train sets are disjoint, union to a full dataset, ect.

File: albatross/core/indexing.h [hunk 2]
```diff
@@ -125,34 +99,39 @@ symmetric_subset(const std::vector<SizeType> &indices,
...
-  } else {
-    return Distribution<CovarianceType>(mean);
+/*
+ * Set a subset of an Eigen::Vector.  If this worked it'd be
+ * the equivalent of:
+ *
+ *     to[indices] = from;
+ */
+template <typename SizeType>
+inline void set_subset(const std::vector<SizeType> &indices,
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/indexing.h]: yeah, so `set_subset(&to, indices, from);`? I like that argument order better to.

File: albatross/evaluation/cross_validation.h [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+  template <typename RequiredPredictType, typename FeatureType>
+  Eigen::VectorXd scores(const EvaluationMetric<RequiredPredictType> &metric,
+                         const RegressionDataset<FeatureType> &dataset,
+                         const GrouperFunction<FeatureType> &grouper) const {
+    const auto indexer = leave_one_group_out_indexer(dataset.features, grouper);
+    return scores(metric, dataset, indexer);
+  }
+
+  template <typename RequiredPredictType, typename FeatureType>
+  Eigen::VectorXd scores(const EvaluationMetric<RequiredPredictType> &metric,
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation.h]: yeah, I could see it being confusing.  As an alternative I could just remove these special case functions and force the user to do:
`scores(metric, dataset, leave_one_out_indexer(dataset.features))`.
which would definitely make it more explicit.

Or yet another option I considered was to completely remove the `get_prediction` part:

`model.cross_validate().leave_one_out(dataset).scores()`

File: albatross/evaluation/cross_validation_utils.h (hunk 1 above)
Comment (akleeman) [reply albatross/evaluation/cross_validation_utils.h]: yeah I guess I should probably do a full check to make sure the folds completely cover the dataset.

File: albatross/evaluation/cross_validation.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation.h]: What do you think about this:
```
  template <typename FeatureType>
  CVPrediction<ModelType, FeatureType>
  leave_out(const RegressionDataset<FeatureType> &dataset,
            const FoldIndexer &indexer) const {
    return CVPrediction<ModelType, FeatureType>(model_, dataset, indexer);
  }

  template <typename FeatureType>
  auto leave_one_group_out(const RegressionDataset<FeatureType> &dataset,
                           const GrouperFunction<FeatureType> &grouper) const {
    const auto indexer = leave_one_group_out_indexer(dataset.features, grouper);
    return leave_out(dataset, indexer);
  }

  template <typename FeatureType>
  auto leave_one_out(const RegressionDataset<FeatureType> &dataset) const {
    const auto indexer = leave_one_out_indexer(dataset.features);
    return leave_out(dataset, indexer);
  }
```
It strays from the syntax of the models but it's more specific for cross validation.
```
Eigen::VectorXd mean = model.get_fit_model(dataset).get_predictions(dataset.features).mean();
Eigen::VectorXd loo_mean = model.cross_validate().leave_one_out(dataset).mean();
```
vs
```
Eigen::VectorXd mean = model.get_fit_model(dataset).get_predictions(dataset.features).mean();
Eigen::VectorXd loo_mean = model.cross_validate().get_predictions(dataset).mean();
```

[the remark this answers is not in the document]
Comment (seth-swiftnav) [reply albatross/evaluation/cross_validation.h]: Personally, I'm just looking for the fold indices and metric to be explicit. Being able to return the prediction and then compute scores is good, as it allows multiple scores to be computed, and also allows for refolding - as we do now. This latest proposal looks good to me.
But `scores(metric, dataset, leave_one_out_indexer(dataset.features))` hides this.

File: albatross/evaluation/cross_validation_utils.h [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  std::size_t n = 0;
+  for (const auto &fold : folds) {
+    for (const auto &idx : fold.test_indices) {
```
[the remark this answers is not in the document]
Comment (jbangelo) [reply albatross/evaluation/cross_validation_utils.h]: Just to clarify, it's still looping and doesn't make the code any faster but it hides away the inner for loop.

File: albatross/core/fit_model.h [hunk 5]
```diff
@@ -23,15 +23,20 @@ template <typename ModelType, typename Fit> class FitModel {
       std::is_move_constructible<Fit>::value,
       "Fit type must be move constructible to avoid unexpected copying.");
 
+  FitModel(const ModelType &model, Fit &fit) = delete;
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/fit_model.h]: Yeah, good catch.

File: albatross/core/indexing.h [hunk 6]
```diff
@@ -125,34 +99,39 @@ symmetric_subset(const std::vector<SizeType> &indices,
...
+                       const Eigen::VectorXd &from, Eigen::VectorXd *to) {
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/indexing.h]: I *think* this aligns with the conventions we use elsewhere, namely if something is going to get modified in place it's a pointer, otherwise it's a const ref.

File: albatross/core/indexing.h [hunk 7]
```diff
@@ -125,34 +99,39 @@ symmetric_subset(const std::vector<SizeType> &indices,
...
-        test_indices(test_indices_){};
-};
+template <typename SizeType, typename Scalar, int Size>
+inline Eigen::VectorXd
+set_subset(const std::vector<SizeType> &indices,
+           const Eigen::DiagonalMatrix<Scalar, Size> &from,
+           Eigen::DiagonalMatrix<Scalar, Size> *to) {
+  assert(static_cast<Eigen::Index>(indices.size()) == from.size());
+  for (std::size_t i = 0; i < indices.size(); i++) {
+    to.diagonal()[static_cast<Eigen::Index>(indices[i])] =
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/indexing.h]: yeah, interesting ... I suppose that function isn't used anywhere.  will add tests.

File: albatross/core/indexing.h [hunk 8]
```diff
@@ -125,34 +99,39 @@ symmetric_subset(const std::vector<SizeType> &indices,
...
+        from.diagonal()[static_cast<Eigen::Index>(i)];
+  }
+}
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/indexing.h]: yeah, function signature is wrong.

File: albatross/core/prediction.h [hunk 9]
```diff
@@ -150,9 +143,28 @@ class Prediction {
   Eigen::VectorXd
   joint() const = delete; // No valid predict joint method found.
 
+  template <typename PredictType>
+  PredictType get(PredictTypeIdentity<PredictType> =
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/prediction.h]: mostly it just makes it so you don't have redundency when actually call this method.  Ie,

    `prediction.get<MarginalDistribution>();`

instead of

    `prediction.get<MarginalDistribution>(PredictTypeIdentity<MarginalDistribution>());`

File: albatross/evaluation/cross_validation_utils.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation_utils.h]: per seth's suggestion I actually got rid of this function in favor of one that get's all the unique indices:
  `std::set<std::size_t> unique_test_indices(const std::vector<RegressionFold<FeatureType>> &folds)`
and another which uses the output of that at which point I can use `max_element`.

File: albatross/evaluation/cross_validation_utils.h [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    const std::vector<RegressionFold<FeatureType>> &folds,
+    const std::vector<Eigen::VectorXd> &means) {
+  assert(folds.size() == means.size());
+
+  Eigen::Index n = static_cast<Eigen::Index>(max_index(folds));
+  Eigen::VectorXd pred(n);
+  Eigen::Index number_filled = 0;
+  // Put all the predicted means back in order.
+  for (std::size_t i = 0; i < folds.size(); ++i) {
+    const auto fold_pred = means[i];
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation_utils.h]: whoa, not intentional.  Definitely wouldn't have thought that was copying.  What is causing the copy?

File: albatross/evaluation/cross_validation_utils.h [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+template <typename FeatureType, typename RequiredPredictType>
+Eigen::VectorXd
+cross_validated_scores(const EvaluationMetric<RequiredPredictType> &metric,
+                       const std::vector<RegressionFold<FeatureType>> &folds,
+                       const std::vector<RequiredPredictType> &predictions) {
+  assert(folds.size() == predictions.size());
+  Eigen::Index n = static_cast<Eigen::Index>(predictions.size());
+  Eigen::VectorXd output(n);
+  for (Eigen::Index i = 0; i < n; ++i) {
```
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation_utils.h]: the output is an `Eigen::VectorXd` is that still compatible with `std::transform`?

File: albatross/core/indexing.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbangelo) [reply albatross/core/indexing.h]: Ah, I missed the bit in our C++ coding standard that said output parameters should be pointers. It makes sense, it usually forces the called to knowingly pass in a pointer.

It also strongly suggests using the return value for output instead of using a parameter when possible. `Eigen::VectorXd` not having efficient move semantics would be a reasonable motivation to not simply return them.

File: albatross/core/indexing.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbangelo) [reply albatross/core/indexing.h]: I'd suggest `set_subset(from, indices, &to);` Our C++ coding standard suggests putting output parameters last in the signature, and that lines up with most of the C/C++ style out there. When passing C style arrays around it's usually the pointer first directly followed by the array length, which is somewhat analogous to our use of `from` and `indicies` here.

File: albatross/core/prediction.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (jbangelo) [reply albatross/core/prediction.h]: I meant why can't we get rid of the argument all together? So you'd still be able to do
```prediction.get<MarginalDistribution>();```

File: albatross/evaluation/cross_validation.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation.h]: Yeah, Re:computing multiple metrics I'm trying to set this all up in a way where if you'd rather you can do everything manually using free functions the way it was before this change.  Ie,

```
auto folds = folds_from_fold_indexer(dataset, indexer);
auto predictions = get_predictions(model, folds);
auto foo_scores = cross_validated_scores(foo_metric, folds, predictions);
auto bar_scores = cross_validated_scores(bar_metric, folds, predictions);
```

Re:naming I'm kind of torn about being super explicit about each variant of computing folds `leave_one_out_scores`, `leave_one_group_out_scores` instead of just having a unified `scores` method whose behavior is (hopefully) obvious from the arguments.  In fact, the more I think about it, that just seems better.  In which case the question is, do we want it to default to leave one out?  Or just not have a default.  FWIW [`scikit-learn` has a default](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.cross_val_score.html#sklearn.model_selection.cross_val_score) which (of course) is 3 folds ?!?.  I suppose that default is designed for dealing with really large datasets.

File: albatross/core/prediction.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/prediction.h]: Ah, yeah I see.  The extra parameter is necessary since the return types are different.

File: albatross/core/indexing.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/core/indexing.h]: As it currently stands I have `subset(indices, from)`, so I'm cool with changing to `set_subset(from, indices, &to)`, but I'll probably change `subset(from, indices)` to match it.  That make sense to you guys?

File: albatross/evaluation/cross_validation_utils.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (jbangelo) [reply albatross/evaluation/cross_validation_utils.h]: `auto` follows the [template type deduction rules from a function call](https://en.cppreference.com/w/cpp/language/template_argument_deduction). The short answer being that any references and cv-qualifiers are stripped off the type before deduction is performed, meaning using a bare `auto` won't get you a reference.

[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation_utils.h]: Crazy!  Good to know.  I suspect we have lots of unintentional copies in our code ...

File: albatross/evaluation/cross_validation_utils.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (jbangelo) [reply albatross/evaluation/cross_validation_utils.h]: `std::transform` operates on iterator like objects. So assuming `Eigen::VectorXd` provides an interator object it would look something like this 

```
template <typename FeatureType, typename RequiredPredictType>
Eigen::VectorXd
cross_validated_scores(const EvaluationMetric<RequiredPredictType> &metric,
                       const std::vector<RegressionFold<FeatureType>> &folds,
                       const std::vector<RequiredPredictType> &predictions) {
  assert(folds.size() == predictions.size());
  Eigen::Index n = static_cast<Eigen::Index>(predictions.size());
  Eigen::VectorXd output(n);
  auto make_metric = [](const auto &fold, const auto& prediction) {
    auto fold_size = static_cast<std::size_t>(fold.test_dataset.size());
    auto pred_size = static_cast<std::size_t>(prediction.size());
    assert(fold_size == pred_size);
    return metric(prediction, fold.test_dataset.targets);
  }
  std::transform(folds.begin(), folds.end(), predictions.begin(), output.begin(), make_metric);
}
```

Like I said, probably doesn't increase readability much.

[the remark this answers is not in the document]
Comment (akleeman) [reply albatross/evaluation/cross_validation_utils.h]: Ah yeah, I was assuming `Eigen::VectorXd` didn't support iterators, but it does as of 3.4: `https://eigen.tuxfamily.org/dox-devel/group__TutorialSTL.html`
